File: server/internal/domain/ranking/contract.go
package ranking

import (
	"github.com/google/uuid"
	"github.com/shopspring/decimal"
)

type ScorePolicy struct {
	Version   string                     `json:"version"`
	MinTrades int                        `json:"min_trades"`
	Weights   map[string]decimal.Decimal `json:"weights"`
}

type LeaderboardEntry struct {
	ID                 uuid.UUID       `json:"entry_id"`
	EvaluationID       uuid.UUID       `json:"evaluation_id"`
	Score              decimal.Decimal `json:"score"`
	Rank               int             `json:"rank"`
	ScorePolicyVersion string          `json:"score_policy_version"`
}

type RankingService interface {
	Rank(ScorePolicy, decimal.Decimal, int) (LeaderboardEntry, error)
}

File: py_backend/app/domain/ranking/contract.py
"""Ranking / leaderboard contracts (float64).

Mirrors `server/internal/domain/ranking/contract.go`.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Protocol
from uuid import UUID

from ..common import Decimal


@dataclass
class ScorePolicy:
    version: str
    min_trades: int
    weights: dict[str, Decimal] = field(default_factory=dict)


@dataclass
class LeaderboardEntry:
    entry_id: UUID
    evaluation_id: UUID
    score: Decimal
    rank: int
    score_policy_version: str


class RankingService(Protocol):
    def rank(self, policy: ScorePolicy, score: Decimal, min_trades: int) -> LeaderboardEntry: ...
